### PR TITLE
Selection set to end on input value change

### PR DESCRIPTION
**Pull request by Matthew-Goldberg**
Fix for https://github.com/jsdom/jsdom/issues/2787

When setting the value of a `textarea` or `input type=text`, the selection should be set to the end of the new text.

Also fixed an issue in `textarea`, the selection start/end should initially be set to 0.

### CODE REVIEW

**Comment by Matthew-Goldberg** [comment]
`set-value-reset-selection.html` Failed in "textarea setting value should reset selection indexes back to defaults"
Seems to me this test is incorrect, the browser does not appear to behave this way.

**Comment by Matthew-Goldberg** [comment]
Yes it appears that `set-value-reset-selection.html` is no longer valid, spec updated in https://github.com/whatwg/html/pull/3472
Test can be updated or deleted.  I think it can be deleted since that test is covered now in `html/semantics/forms/textfieldselection`

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js [hunk 1]
```diff
@@ -69,12 +71,14 @@ class HTMLTextAreaElementImpl extends HTMLElementImpl {
...
 
   set value(val) {
+    const oldValue = this._rawValue;
     this._rawValue = val;
     this._dirtyValue = true;
 
-    this._selectionStart = 0;
-    this._selectionEnd = 0;
-    this._selectionDirection = "none";
+    if (oldValue !== this._rawValue) {
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: This doesn't seem to match the spec; see https://html.spec.whatwg.org/multipage/form-elements.html#dom-textarea-value. You need to use API value instead of raw value.

File: test/web-platform-tests/to-run.yaml [hunk 2]
```diff
@@ -574,6 +574,17 @@ DIR: html/semantics/forms/resetting-a-form
...
 
+DIR: html/semantics/forms/textfieldselection
+
+select-event.html: [fail, not yet implemented]
+selection-after-content-change.html: [fail, not yet implemented]
+selection-not-application-textarea.html: [fail, not yet implemented]
+selection-start-end-extra.html: [fail, not yet implemented]
+textfieldselection-setRangeText.html: [fail, not yet implemented]
+textfieldselection-setSelectionRange.html: [fail, not yet implemented]
+
```
Comment (domenic) [inline test/web-platform-tests/to-run.yaml]: Which tests in this directory fail before your change but pass afterward?

Comment (Matthew-Goldberg) [reply test/web-platform-tests/to-run.yaml]: The following now pass:
```
selection-not-application.html
selection-start-end.html
selection-value-interactions.html
selection.html
setSelectionRange.html
textarea-selection-while-parsing.xhtml
```

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js (hunk 1 above)
Comment (Matthew-Goldberg) [reply lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: Thank you this is very helpful.  Looks like this may fix an additional test as well.  Working on it.

Comment (Matthew-Goldberg) [reply lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: This has been fixed, and now got `selection-after-content-change` to pass.

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js [hunk 3]
```diff
@@ -69,12 +76,14 @@ class HTMLTextAreaElementImpl extends HTMLElementImpl {
   }
 
   set value(val) {
+    const oldValue = this._getValue();
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: This still doesn't seem to match the spec, which uses API value (obtained in jsdom via `_getAPIValue()`).

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js [hunk 4]
```diff
@@ -44,7 +46,12 @@ class HTMLTextAreaElementImpl extends HTMLElementImpl {
     super._childTextContentChangeSteps();
 
     if (this._dirtyValue === false) {
+      const oldValue = this._getValue();
       this._rawValue = childTextContent(this);
+      if (oldValue !== this._getValue()) {
+        this._selectionStart = this._selectionEnd = 0;
+        this._selectionDirection = "none";
+      }
```
Comment (domenic) [inline lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: I don't see this in the spec. Is this change necessary to pass any tests?

File: test/web-platform-tests/to-run.yaml (hunk 2 above)
Comment (Matthew-Goldberg) [reply test/web-platform-tests/to-run.yaml]: `selection-after-content-change.html` also passes after the change to use `_getAPIValue` to match spec.

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js (hunk 3 above)
Comment (Matthew-Goldberg) [reply lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: Fixed

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js (hunk 4 above)
Comment (Matthew-Goldberg) [reply lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: You are right, this is not needed for these tests.  I have removed this part of the change.

File: lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js [hunk 5]
```diff
@@ -69,12 +71,14 @@ class HTMLTextAreaElementImpl extends HTMLElementImpl {
   }
 
   set value(val) {
+    const oldValue = this._getAPIValue();
```
Comment (TimothyGu) [inline lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: nit: the spec calls it _oldAPIValue_.

Comment (Matthew-Goldberg) [reply lib/jsdom/living/nodes/HTMLTextAreaElement-impl.js]: Thanks, I renamed this to `oldAPIValue` and added a comment linking to the spec.
